src/scheduler/MpiThreadPool.cpp

```diff
@@ -0,0 +1,98 @@
+#include <faabric/scheduler/MpiThreadPool.h>
+#include <faabric/util/logging.h>
+
+namespace faabric::scheduler {
+MpiAsyncThreadPool::MpiAsyncThreadPool(int nThreads)
+  : shutdown(false)
+{
+    faabric::util::getLogger()->debug(
+      "Starting an MpiAsyncThreadPool of size {}", nThreads);
+
+    // Initialize job queue
+    localJobQueue = std::make_shared<MpiJobQueue>();
+
+    // Initialize thread pool
+    for (int i = 0; i < nThreads; ++i) {
+        threadPool.emplace_back(
+          std::bind(&MpiAsyncThreadPool::entrypoint, this, i));
+    }
+}
+
+MpiAsyncThreadPool::~MpiAsyncThreadPool()
+{
+
+    faabric::util::getLogger()->debug("Shutting down MpiAsyncThreadPool");
+    this->shutdown = true;
+
+    // When the destructor is called, either some thread will have hit the
+    // dequeue timeout and won't be joinable, or none has and all are joinable.
+    // We wait untill all hit the timeout, saving an additional cond. variable.
+    for (auto& thread : threadPool) {
+        if (thread.joinable()) {
+            thread.join();
+        }
+    }
+}
+
+void MpiAsyncThreadPool::awaitAsyncRequest(int reqId)
```

Comment:
This implements the logic that before would have been in `faabric::awaitAsyncRequest`. I've kept both as I like the indirection: `faasm->faabric->threadPool_impl`